presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java

```diff
@@ -1720,6 +1757,26 @@ else if (criteria instanceof JoinOn) {
                     }
                 }
 
+                if (!onConditionReferencesRelatedFields(expression, node.getRight())
+                        && (node.getRight() instanceof AliasedRelation || node.getRight() instanceof Table)) {
```

Comment:
I am not sure why are we checking only for AliasedRelation or Table type only. The right node very well could be anything. In case of subquery this logic would not work for a case like
select * from t1 inner join ( select t2.id1 as t2_id1, t3.id1 from t2 inner join t3 on t2.id1=t3.id1) ON t1.id1 = t1.id1;

@kaikalur Any reason we would not want to handle nested joins?

Reply:
Hmm, that should be caught by the current logic, I think. The main idea here is to catch simple copy/paste bugs. We will need another deeper check in the planner where we don't find predicates involving all tables of the join.,

Reply:
This change will not handle TableSubquery as it needs to be known the set of columns associated with that relation. There is no such information at this level.

This example is similar to the one  that Sreeni pointed out earlier `select t1 join t2 on x1=y1 join t3 on x2=y2`